crates/apollo-router-core/src/error.rs

```diff
@@ -185,6 +186,44 @@ impl From<JoinError> for QueryPlannerError {
 pub enum SchemaError {
     /// IO error: {0}
     IoError(#[from] std::io::Error),
-    /// Parsing error(s).
-    ParseErrors(Vec<apollo_parser::Error>),
+    /// Parsing error(s): {0}
+    Parse(ParseErrors),
+}
+
+#[derive(Debug)]
+pub struct ParseErrors {
+    raw_schema: String,
+    errors: Vec<apollo_parser::Error>,
+}
+
+impl SchemaError {
+    pub fn from_parse_errors(raw_schema: String, errors: Vec<apollo_parser::Error>) -> Self {
+        Self::Parse(ParseErrors { raw_schema, errors })
+    }
+}
+
+#[derive(Error, Debug, Diagnostic)]
+#[error("{}", self.ty)]
+#[diagnostic(code("apollo-parser parsing error."))]
+struct ParserError {
+    ty: String,
+    #[source_code]
+    src: NamedSource,
+    #[label("{}", self.ty)]
+    span: SourceSpan,
+}
+
+impl std::fmt::Display for ParseErrors {
```

Comment:
Hmm so what you mean is that we're conflating Error propagation and reporting, which is very true!

I wonder what responsibility should lie within `Display` and to the std::fmt::formatter. I would for example have liked to be able to ask the formatter if it supports colors.

Reply:
There might be a couple of clues https://docs.rs/miette/3.2.0/miette/trait.ReportHandler.html around here, I'll let you know what I've found!

Reply:
> I would for example have liked to be able to ask the formatter if it supports colors.

afaik this is unsupported. That is why I think you can't put ansi code in implementations of Debug and Display.

For example ToString depends on Display. Is a string supposed to have ansi codes like that?

But then really my concern is that we report errors through other channels (journald, syslog, etc...) and we incorrectly send ansi codes in them.

Again this is all fancy but we should make it clear that this targets only the terminal (aka stdout or stderr if and only if it is a tty, otherwise you need to strip the ansi codes).

Reply:
ohhh i did not know this is how you were going to send the error to the logs! I think the logs should just get the error slice, nothing should be formatted at all! Whatever reads the logs can do the formatting, or you can pipe the logs to a formatter.

Reply:
Well it's in flux really, We do show the logs to a console at the moment, so we might as well have nice ones. OTOH we do have log exporters as well, so I'll probably move the formating to the "console logging layer", I feel this is what @cecton is hinting at?

Reply:
To be sure we understand each other:

1. Remove the `{0}` from the doc comment which will prevent this enum's Display to use the Display impl of ParseErrors. If the error gets logged, the user will have no other message than "Parsing error(s).". This is **fine**.
    ```rust
    /// Parsing error(s).
    Parse(ParseErrors),
    ```

2.  Change this piece of code:
    ```rust
    impl std::fmt::Display for ParseErrors {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            // ...
        }
    }
    ```
    To this:
    ```rust
    impl ParseErrors {
        fn print_pretty(&self) {
            // ... also replace writeln!(f, ...) by eprintln!(...)
        }
    }
    ```
    So we don't end up with a Display implementation and a ToString implementation that renders ansi codes (things like this `ESC[38;5;` in the string. See [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code)).

3.  In `crates/apollo-router-core/src/schema.rs`, instead of just simply returning the error, use your `print_pretty` function: 
    ```rust
    if !tree.errors().is_empty() {
        let err = SchemaError::from_parse_errors(
            s.to_string(),
            tree.errors().to_vec(),
        );
        if isatty() {
            err.print_pretty();
        }
        return Err(err);
    ```
    You'll need this: [isatty](https://github.com/dtolnay/isatty)